fixes for gplay store after 2019-09-01 updates

icarcal opened the pull request:
This PR fixes the following issues: 
#333 
#334 
#335 
#336 

I have updated the following methods:
 ```.list```
 ```.reviews```
 ```.permissions```
 ```.categories```

 ```.list ``` and  ```.reviews ``` methods introduce some breaking changes, since google allow you to fetch data only through ```batchexecute``` endpoint.
You can now only send the ```num``` parameter (it will allow you to fetch the desired number of category apps and reviews).
The ```.list``` method is unstable right now, since google sometimes respond with 200 apps, and sometimes more (for example 600 and so on).

For the ```.permissions``` method I've added the ```type``` parameter to the response (since this is not a breaking change but add value to it).
For the ```.reviews``` method I've added the ```thumbsUp``` and ```version``` parameters, this is also not a breaking change.

I've also updated the tests and docs

abubakaraziz commented:
@icarcal There are some issues with Permissions method. UnhandledPromiseRejectionWarning: TypeError: html[0].flatMap is not a function.
Please look at the screenshot attached. 
![Screenshot from 2019-09-09 11-12-36](https://user-images.githubusercontent.com/10674679/64507061-16a04980-d2f3-11e9-820e-99302b54a51b.png)

yurganov commented:
Please increase numberOfApps to 100 in appsMappedRequests.js

yurganov commented:
Great work!

facundoolano commented on README.md:
```diff
@@ -290,7 +290,7 @@ Results:
     scoreText: '5',
     url: 'https://play.google.com/store/apps/details?id=com.dxco.pandavszombies&reviewId=Z3A6QU9xcFRPRWZaVHVZZ081NlNsRW9TV0hJeklGSTBvYTBTUlFQUUJIZThBSGJDX2s1Y1o0ZXRCbUtLZmgzTE1PMUttRmpRSS1YcFgxRmx1ZXNtVzlVS0Zz'
-    title: 'I LOVE IT',
+    title: 'I LOVE IT', //GOOGLE DO NOT RETURN THE TITLE ANYMORE
```
then remove it from the documentation

icarcal replied:
I'm still returning this as ```null```, otherwise it will be another breaking change.
Should I remove or just change the documentation?

facundoolano commented:
Great work, thanks for this. Since you are fixing/improving separate issues, would you mind issuing separate PRs for the changes that are not related? This is a very big diff, it gets difficult to review.

icarcal commented:
> Great work, thanks for this. Since you are fixing/improving separate issues, would you mind issuing separate PRs for the changes that are not related? This is a very big diff, it gets difficult to review.

Of course, going to separate the changes into different PRs.
I'll let this one open untill I separate the PRs, is it ok?

marxsk commented:
@icarcal There is an issue with reviews, it returns reviews that I cannot see on the web :/ and it does not show those that are shown. I'm happy to help with debugging / patching if you have an idea what might goes wrong.

```
  const options = {
    sort: gplay.sort.NEWEST,
    num: 100,
    appId: 'foo',
    lang: 'cz',
  };
  return gplay.reviews(options);
```

icarcal commented:
@marxsk can you provide at least one appId and your country code?
I probably missed the gl parameter.

icarcal commented:
@abubakaraziz you are just running the tests and getting this error?
 What node version are you running?

The [TravisCI](https://travis-ci.org/facundoolano/google-play-scraper/builds/582684119) passes this tests and it is using node version 12.

Going to use another method since flatMap was introduced at version [11.15](https://node.green), for  backward compatibility

abubakaraziz commented:
@icarcal  I am using node version 10 so I think that's why it's causing the issue.

marxsk commented:
@icarcal sure, this one does not have tons of revies - cz.ulikeit.fio 

Be aware of issue: https://github.com/facundoolano/google-play-scraper/issues/276 - with geo IP; so it might not work for you as expected.

icarcal commented:
@abubakaraziz do you mind testing it again?
Changed from ES2019 flatMap to ramda chain, which is the same.

@umuero thanks for pointing this!
Gonna take a better look at pagination.

@marxsk apparently now we can use the gl parameter for reviews :tada:, I have added the country parameter, do you mind checking that again?

marxsk commented:
@icarcal testing done - it does not work but we are close :)

It looks like that my problem is not 'gl' but 'hl' this looks like the main difference between chrome request and yours PR.

icarcal commented:
@umuero fixed the pagination for reviews!
I forgot to pass a parameter to the getBody function, so it was getting only the first page over and over.
It look ok now.

abubakaraziz commented:
@icarcal gPlay.permissions is working fine now, but permission and description are fetching exactly the same information.

icarcal commented:
> @icarcal gPlay.permissions is working fine now, but permission and description are fetching exactly the same information.

google does not provide different infos anymore, as a backward compatibility I left the exact same description and permission, as show at the store.

The response we are getting from google right now is:
```
[
    [
        [
            "Photos/Media/Files",
            [
                null,
                2,
                null,
                [
                    null,
                    null,
                    "https://lh3.googleusercontent.com/pHtIujPWxciAZcfYSwlrGGq14Z984rKLMgcm9RPATLiOlbrWy-tVlelEWgED7gpktgcD1tZizVeHiO5fkw"
                ]
            ],
            [
                [
                    null,
                    "read the contents of your USB storage"
                ],
                [
                    null,
                    "modify or delete the contents of your USB storage"
                ]
            ]
        ],
        [
            "Storage",
            [
                null,
                2,
                null,
                [
                    null,
                    null,
                    "https://lh3.googleusercontent.com/aWNKQedLTpw6u6yyMjQObmuoKu67A1czWnIcvID86oAmMT02r5mNdRn6l9ZN2t2MIyH6tNy-01v7ukeQ"
                ]
            ],
            [
                [
                    null,
                    "read the contents of your USB storage"
                ],
                [
                    null,
                    "modify or delete the contents of your USB storage"
                ]
            ]
        ],
        [
            "Microphone",
            [
                null,
                2,
                null,
                [
                    null,
                    null,
                    "https://lh3.googleusercontent.com/daUjqbSOr2QpaqXS2HQbNzYzzqN2yWGzM_7AZxwFaWLT7_kIhX95HKi_HSpjeeQDOmFMENZxJqblbu_4qg"
                ]
            ],
            [
                [
                    null,
                    "record audio"
                ]
            ]
        ]
    ],
    [
        [
            "Other",
            [
                null,
                2,
                null,
                [
                    null,
                    null,
                    "https://lh3.googleusercontent.com/pkKXoPl5q7n8T0s7KREtdvUZn1PLRgx-Ox0t4tkO8af4JpgGbyAxLBTsvEKKBCjwBACQsZisSYNmHPGbBA"
                ]
            ],
            [
                [
                    null,
                    "full network access"
                ],
                [
                    null,
                    "prevent device from sleeping"
                ],
                [
                    null,
                    "read Google service configuration"
                ],
                [
                    null,
                    "view network connections"
                ]
            ]
        ]
    ],
    [
        [
            null,
            "interact across users"
        ],
        [
            null,
            "receive data from Internet"
        ]
    ]
]```

It is the same as shown in the store:
![Screenshot from 2019-09-09 12-31-56](https://user-images.githubusercontent.com/11530480/64544706-d72b2a80-d2fd-11e9-95b7-86c35f88801f.png)

abubakaraziz commented:
@icarcal Yes, I have also checked. It doesn't provide descriptions anymore on the Playstore as well.

marxsk commented:
@icarcal It works for me - I was the problem. I have replaced 'lang' with 'country'. When I have both of them set it works as expected. Thanks for a patch.

gamechosun commented:
@icarcal Thank you! This fixes works well.

icarcal commented:
> would you mind issuing separate PRs for the changes that are not related? This is a very big diff, it gets difficult to review.

@facundoolano I've splitted this PR into 4 different ones as requested:
#338 
#339 
#340 
#341

facundoolano commented:
closing this as the code has been splitted in other PRs

facundoolano commented:
Merged the other PRs and published a new major version. Thanks for all the hard work @icarcal !
